// @stimulus/core/dist/module/test/log_controller.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Controller } from "stimulus";
var LogController = /** @class */ (function (_super) {
    __extends(LogController, _super);
    function LogController() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.initializeCount = 0;
        _this.connectCount = 0;
        _this.disconnectCount = 0;
        _this.actionLog = [];
        return _this;
    }
    LogController.bless = function () {
        _super.bless.call(this);
        this.blessCount++;
    };
    LogController.prototype.initialize = function () {
        this.initializeCount++;
    };
    LogController.prototype.connect = function () {
        this.connectCount++;
    };
    LogController.prototype.disconnect = function () {
        this.disconnectCount++;
    };
    LogController.prototype.log = function (event) {
        this.recordAction("log", event);
    };
    LogController.prototype.log2 = function (event) {
        this.recordAction("log2", event);
    };
    LogController.prototype.recordAction = function (name, event) {
        var entry = { name: name, eventType: event.type, eventTarget: event.target, defaultPrevented: event.defaultPrevented };
        this.actionLog.push(entry);
    };
    LogController.blessCount = 0;
    return LogController;
}(Controller));
export { LogController };
