// @stimulus/core/dist/module/test/cases/lifecycle_tests.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { LogControllerTestCase } from "../log_controller_test_case";
var LifecycleTests = /** @class */ (function (_super) {
    __extends(LifecycleTests, _super);
    function LifecycleTests() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LifecycleTests.prototype.setup = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.controllerElement = this.controller.element;
                return [2 /*return*/];
            });
        });
    };
    LifecycleTests.prototype["test Controller#initialize"] = function () {
        return __awaiter(this, void 0, void 0, function () {
            var controller;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        controller = this.controller;
                        this.assert.equal(controller.initializeCount, 1);
                        return [4 /*yield*/, this.reconnectControllerElement()];
                    case 1:
                        _a.sent();
                        this.assert.equal(this.controller, controller);
                        this.assert.equal(controller.initializeCount, 1);
                        return [2 /*return*/];
                }
            });
        });
    };
    LifecycleTests.prototype["test Controller#connect"] = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.assert.equal(this.controller.connectCount, 1);
                        return [4 /*yield*/, this.reconnectControllerElement()];
                    case 1:
                        _a.sent();
                        this.assert.equal(this.controller.connectCount, 2);
                        this.assert.equal(this.controller.constructor["blessCount"], 1);
                        return [2 /*return*/];
                }
            });
        });
    };
    LifecycleTests.prototype["test Controller#disconnect"] = function () {
        return __awaiter(this, void 0, void 0, function () {
            var controller;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        controller = this.controller;
                        this.assert.equal(controller.disconnectCount, 0);
                        return [4 /*yield*/, this.disconnectControllerElement()];
                    case 1:
                        _a.sent();
                        this.assert.equal(controller.disconnectCount, 1);
                        return [2 /*return*/];
                }
            });
        });
    };
    LifecycleTests.prototype.reconnectControllerElement = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.disconnectControllerElement()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.connectControllerElement()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    LifecycleTests.prototype.connectControllerElement = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.fixtureElement.appendChild(this.controllerElement);
                        return [4 /*yield*/, this.nextFrame];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    LifecycleTests.prototype.disconnectControllerElement = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.fixtureElement.removeChild(this.controllerElement);
                        return [4 /*yield*/, this.nextFrame];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return LifecycleTests;
}(LogControllerTestCase));
export default LifecycleTests;
