// @stimulus/mutation-observers/dist/module/test/lib/delegate_recorders.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var Recorder = /** @class */ (function () {
    function Recorder() {
        this.reset();
    }
    Object.defineProperty(Recorder.prototype, "lastEntry", {
        get: function () {
            return this.entries[this.entries.length - 1];
        },
        enumerable: true,
        configurable: true
    });
    Recorder.prototype.record = function (methodName, args) {
        var entry = {};
        entry[methodName] = args;
        this.entries.push(entry);
    };
    Recorder.prototype.reset = function () {
        this.entries = new Array;
    };
    return Recorder;
}());
export { Recorder };
var AttributeObserverRecorder = /** @class */ (function (_super) {
    __extends(AttributeObserverRecorder, _super);
    function AttributeObserverRecorder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AttributeObserverRecorder.prototype.elementMatchedAttribute = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.record("elementMatchedAttribute", args);
    };
    AttributeObserverRecorder.prototype.elementAttributeValueChanged = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.record("elementAttributeValueChanged", args);
    };
    AttributeObserverRecorder.prototype.elementUnmatchedAttribute = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.record("elementUnmatchedAttribute", args);
    };
    return AttributeObserverRecorder;
}(Recorder));
export { AttributeObserverRecorder };
var TokenListObserverRecorder = /** @class */ (function (_super) {
    __extends(TokenListObserverRecorder, _super);
    function TokenListObserverRecorder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TokenListObserverRecorder.prototype.elementMatchedTokenForAttribute = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.record("elementMatchedTokenForAttribute", args);
    };
    TokenListObserverRecorder.prototype.elementUnmatchedTokenForAttribute = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.record("elementUnmatchedTokenForAttribute", args);
    };
    return TokenListObserverRecorder;
}(Recorder));
export { TokenListObserverRecorder };
