// @stimulus/mutation-observers/dist/module/test/lib/environment.js
import { setFixture } from "./helpers";
var TestEnvironment = /** @class */ (function () {
    function TestEnvironment(observerConstructor, recorderConstructor) {
        this.element = document.createElement("div");
        this.childElement = document.createElement("span");
        this.attributeName = "data-test";
        this.recorder = new recorderConstructor;
        this.observer = new observerConstructor(this.element, this.attributeName, this.recorder);
    }
    TestEnvironment.setup = function (observerConstructor, recorderConstructor) {
        var environment = new TestEnvironment(observerConstructor, recorderConstructor);
        environment.setup();
        return environment;
    };
    TestEnvironment.prototype.setup = function () {
        this.element.appendChild(this.childElement);
        setFixture(this.element);
        this.observer.start();
    };
    TestEnvironment.prototype.teardown = function () {
        this.observer.stop();
        setFixture("");
    };
    return TestEnvironment;
}());
export { TestEnvironment };
